ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs

```diff
@@ -27,24 +64,281 @@ tests =
+                       _ -> Nothing
+                )
+            . selectDiffusionPeerSelectionEvents
+            $ events
+
+          -- Left () is a TrJoiningNetwork event
+          -- Right () is a TrKillingNode event
+          trJoinKillSig :: Signal (Either () ())
+          trJoinKillSig =
+              Signal.fromChangeEvents (Right ()) -- Default to TrKillingNode
+            . Signal.selectEvents
+                (\case TrJoiningNetwork -> Just (Left ())
+                       TrKillingNode    -> Just (Right ())
+                       _                -> Nothing
+                )
+            . selectDiffusionSimulationTrace
+            $ events
+
+          -- Signal.keyedUntil receives 2 functions one that sets start of the
+          -- set signal, one that ends it and another that stops all.
+          --
+          -- In this particular case we want a signal that is keyed beginning
+          -- on a TrJoiningNetwork and ends on TrKillingNode, giving us a Signal
+          -- with the periods when a node was alive.
+          trIsNodeAlive :: Signal Bool
+          trIsNodeAlive =
+                not . Set.null
+            <$> Signal.keyedUntil (either Set.singleton (const Set.empty))
+                                  (either (const Set.empty) Set.singleton)
+                                  (const False)
+                                  trJoinKillSig
+
+          promotionOpportunities :: Signal (Set NtNAddr)
+          promotionOpportunities =
+            (\local established recentFailures inProgressPromoteCold isAlive ->
+              if isAlive
+              then local Set.\\ established
+                         Set.\\ recentFailures
+                         Set.\\ inProgressPromoteCold
+              else Set.empty
+            ) <$> govLocalRootPeersSig
+              <*> govEstablishedPeersSig
+              <*> govEstablishedFailuresSig
+              <*> govInProgressPromoteColdSig
+              <*> trIsNodeAlive
+
+          promotionOpportunitiesIgnoredTooLong :: Signal (Set NtNAddr)
+          promotionOpportunitiesIgnoredTooLong =
+            Signal.keyedTimeout
+              10 -- seconds
+              id
+              promotionOpportunities
+
+       in counterexample
+            ("\nSignal key: (local root peers, established peers, " ++
+             "recent failures, is alive, opportunities, ignored too long)\n" ++
+               intercalate "\n" (map show $ eventsToList events)
+            )
+        $ signalProperty 20 show
+              (\(_,_,_,_,_,_, tooLong) -> Set.null tooLong)
+              ((,,,,,,) <$> govLocalRootPeersSig
+                      <*> govEstablishedPeersSig
+                      <*> govEstablishedFailuresSig
+                      <*> govInProgressPromoteColdSig
+                      <*> trIsNodeAlive
+                      <*> promotionOpportunities
+                      <*> promotionOpportunitiesIgnoredTooLong
```

Comment:
```suggestion
              ((,,,,,,) <$> govLocalRootPeersSig
                        <*> govEstablishedPeersSig
                        <*> govEstablishedFailuresSig
                        <*> govInProgressPromoteColdSig
                        <*> trIsNodeAlive
                        <*> promotionOpportunities
                        <*> promotionOpportunitiesIgnoredTooLong
```

Reply:
Yes we can, however this way we get a better error log message and can easily see what happened.